exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/MongoBinaryValueParser.java

```diff
@@ -0,0 +1,150 @@
+        "(\"<payload>\", \"$type\": \"<t>\") }";
+
+  public MongoBinaryValueParser(JsonStructureParser structParser, ScalarListener listener) {
+    super(structParser, listener);
+  }
+
+  @Override
+  protected String typeName() { return ExtendedTypeNames.BINARY; }
+
+  @Override
+  public void parse(TokenIterator tokenizer) {
+
+    // Null: assume the value is null
+    // (Extension to extended types)
+    JsonToken token = tokenizer.requireNext();
+    if (token == JsonToken.VALUE_NULL) {
+      listener.onValue(token, tokenizer);
+      return;
+    }
+
+    // Value is a scalar, assume binary value as a string.
+    // This is a harmless extension to the standard.
+     if (token.isScalarValue()) {
+      listener.onValue(token, tokenizer);
+      return;
+    }
+
+    // Must be an object
+    requireToken(token, JsonToken.START_OBJECT);
+
+    // { ^($binary | $type)
+    requireToken(tokenizer, JsonToken.FIELD_NAME);
+    String fieldName = tokenizer.textValue();
+    if (fieldName.equals(ExtendedTypeNames.BINARY_TYPE)) {
+      // { $type ^
+      parseV1Format(fieldName, tokenizer);
+    } else if (!fieldName.equals(ExtendedTypeNames.BINARY)) {
+      throw syntaxError();
+    } else if (tokenizer.peek() == JsonToken.START_OBJECT) {
+      // { $binary: ^{
+      parseV2Format(tokenizer);
+    } else {
+      // { $binary: ^value
+      parseV1Format(fieldName, tokenizer);
+    }
+  }
+
+  // Parse field: { ($binary | $type) ^ ...
+  private void parseV1Format(String fieldName, TokenIterator tokenizer) {
+    boolean sawData = false;
+    for (;;) {
+      // key: ^value
+      JsonToken token = requireScalar(tokenizer);
+      if (fieldName.equals(ExtendedTypeNames.BINARY)) {
+        if (sawData) {
+          syntaxError();
+        }
+        sawData = true;
+        listener.onValue(token, tokenizer);
+      }
+
+      // key: value ^(} | key ...)
+      token = tokenizer.requireNext();
+      if (token == JsonToken.END_OBJECT) {
+        break;
+      } if (token != JsonToken.FIELD_NAME) {
+        syntaxError();
+      }
+      fieldName = tokenizer.textValue();
+      switch (fieldName) {
+      case ExtendedTypeNames.BINARY:
+      case ExtendedTypeNames.BINARY_TYPE:
+        break;
+      default:
+        syntaxError();
+      }
+    }
+    if (!sawData) {
+      syntaxError();
+    }
+  }
+
+  // Parse field: { $binary: ^{ "base64": "<payload>", "subType": "<t>" } }
+  // With fields in either order
+  private void parseV2Format(TokenIterator tokenizer) {
+    boolean sawData = false;
+    requireToken(tokenizer, JsonToken.START_OBJECT);
+    for (;;) {
+      JsonToken token = tokenizer.requireNext();
+      if (token == JsonToken.END_OBJECT) {
+        break;
+      } else if (token != JsonToken.FIELD_NAME) {
+        throw syntaxError();
+      }
+      switch (tokenizer.textValue()) {
+        case "base64":
+          if (sawData) {
+            syntaxError();
```

Comment:
Looks like `throw` is missed here and in other places.
```suggestion
            throw syntaxError();
```